src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -194,13 +196,19 @@ internal protected CompletionProvider GetProvider(CompletionItem item)
             {
                 lock (_gate)
                 {
-                    _nameToProvider.TryGetValue(name, out provider);
+                    provider = _nameToProvider.GetOrAdd(name, _getProviderByName);
                 }
             }
 
             return provider;
         }
 
+        private CompletionProvider GetProviderByName(string providerName)
+        {
+            var providers = GetAllProviders(roles: ImmutableHashSet<string>.Empty);
```

Comment:
a comment here explaining why this is the right set of roles would be good.  i actually do not know why this is correct.